File: 프로그래머스/level2/42583.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge = deque()
    bridge_sum = 0
    truck_weights = deque(truck_weights)
    while truck_weights or bridge:
        if bridge and bridge[0][1] >= bridge_length:
            bridge_sum -= bridge[0][0]
            bridge.popleft()
        if truck_weights and bridge_sum + truck_weights[0] <= weight:
            bridge_sum += truck_weights[0]
            bridge.append([truck_weights.popleft(), 0])
        if bridge:
            for truck in bridge:
                truck[1] += 1
        answer += 1
    return answer

print(solution(2, 10, [7, 4, 5, 6]))

File: 프로그래머스/level2/42583.js
function solution(bridgeLength, weight, truckWeights) {
  let truckOnBridge = []
  let bridgeWeight = 0
  let answer = 0

  while (true) {
    if (truckOnBridge.length) {
      if (truckOnBridge[0][0] >= bridgeLength) {
        bridgeWeight -= truckOnBridge.shift()[1]
      }
    }

    if (bridgeWeight + truckWeights[0] <= weight) {
      let truck = truckWeights.shift()
      bridgeWeight += truck
      truckOnBridge.push([0, truck])
    }
    if (!truckWeights.length) {
      return answer + bridgeLength + 1
    }

    truckOnBridge = truckOnBridge.map(t => [t[0] + 1, t[1]])
    answer++
  }
}

console.log(solution(2, 10, [7, 4, 5, 6]))
